crypto/crypto_aesctr_aesni.c

```diff
@@ -26,38 +27,62 @@
  */
 #include "crypto_aesctr_shared.c"
 
-/* Process one whole block of generating & using a cipherblock. */
+/* Process multiple whole blocks by generating & using a cipherblock. */
 static void
-crypto_aesctr_aesni_stream_wholeblock(struct crypto_aesctr * stream,
-    const uint8_t ** inbuf, uint8_t ** outbuf, size_t * buflen_p)
+crypto_aesctr_aesni_stream_wholeblocks(struct crypto_aesctr * stream,
+    const uint8_t ** inbuf, uint8_t ** outbuf, size_t * buflen)
 {
 	__m128i bufsse;
 	__m128i inbufsse;
+	__m128i nonce_be;
+	uint8_t block_counter_be_arr[8];
+	uint64_t block_counter;
+	size_t num_blocks;
+	size_t i;
 
-	/* Prepare counter. */
-	stream->pblk[15]++;
-	if (stream->pblk[15] == 0) {
-		/*
-		 * If incrementing the least significant byte resulted in it
-		 * wrapping, re-encode the complete 64-bit value.
-		 */
-		be64enc(stream->pblk + 8, stream->bytectr / 16);
-	}
+	/* Load local variables from stream. */
+	nonce_be = _mm_loadu_si64(stream->pblk);
+	memcpy(block_counter_be_arr, stream->pblk + 8, 8);
+	block_counter = stream->bytectr / 16;
+
+	/* How many blocks should we process? */
+	num_blocks = (*buflen) / 16;
+
+	for (i = 0; i < num_blocks; i++) {
+		/* Prepare counter. */
+		block_counter_be_arr[7]++;
+		if (block_counter_be_arr[7] == 0) {
+			/*
+			 * If incrementing the least significant byte resulted
+			 * in it wrapping, re-encode the complete 64-bit
+			 * value.
+			 */
+			be64enc(block_counter_be_arr, block_counter);
```

Comment:
This may be a dumb question, but is there a way to use SSE instructions to get a 64-bit value into a register in big-endian format?

Reply:
Yes:
```
static inline __m128i
_mm_bswaplo_epi64(__m128i x)
{
	/* Swap bytes in each 16-bit word. */
        __m128i a = _mm_or_si128(
		_mm_slli_epi16(x, 8),
		_mm_srli_epi16(x, 8));

	/* Reverse 16-bit words in the lowest qword. */
	a = _mm_shufflelo_epi16(a, _MM_SHUFFLE(0, 1, 2, 3));

	return (a);
}
```

(in case anybody lands here from google: this is a personal version that only byte-swaps the lower qword and gives garbage in the upper qword, because that's all we need in aesctr.  To make it work for the entire 128 bits, just add a `_mm_shufflehi_epi16()` in there.)

That's one of the commits that might come in the later PR (in my `-exploratory` branch).  IIRC it gives a +5% bonus to `clang` and a -10% penalty to `gcc`.  (Or it might be the commit that gives a +10% bonus to everything other than `clang -O3 -march=native`, which gets a -40% penalty (!).  I've got commits like that for both sha256 and aesctr.)